In [5]:
# Write a simple Funtion
# Define shout with the parameter, word定义的一个简单function
def shout(word):
    """Print a string with three exclamation marks"""
    # Concatenate the strings: shout_word
    shout_word = word + '!!!'

    # Print shout_word
    print(shout_word)#Returning values is generally more desirable,a print() call assigned to a variable has type NoneType

# 定义simple Funtion




Function的几个组成部分：

Function Header： 
- def + function名字 +（变量1， 变量2）：#最后要有冒号
- def后面四个空格“写入描述这个function”
- Function Body
- return（最后要有）再加上这个function最后会产生的Value
        
        

In [6]:
shout('abc')#加入一个文字string，记得要加入单引号

abc!!!


In [7]:
#Write a single parameter function
#改进上一个shout()function
def shout(word):
    """Return a string with three exclamation marks"""
    # Concatenate the strings: shout_word
    shout_word=word+"!!!"

    # Replace print with return
    return(shout_word)#shout actually returns a value, it does not merely print one.
yell=shout('congratulations')
print(yell)

congratulations!!!


# Multiple parameters and return values
Tuples: 
- 1)like a list, can contain multiple values, 
- 2)immutable, can not modofy valuses,
- 3)construct using圆括号parenthese()



In [12]:
#Write Multiple parameters and return values
#Tuples: 1)like a list, can contain multiple values, 2)immutable, can not modofy valuses, 3)construct using圆括号parenthese()
# Define shout with parameters word1 and word2
def shout(word1, word2):
    """Concatenate strings with three exclamation marks"""
    # Concatenate word1 with '!!!': shout1
    shout1 = word1 + '!!!'
    
    # Concatenate word2 with '!!!': shout2
    shout2 = word2 + '!!!'
    
    # C^oncatenate shout1 with shout2: new_shout
    new_shout = shout1 + shout2#注意这里不需要加括号，将两个value转化成了一个要return的value

    # Return new_shout
    return new_shout

# Pass 'congratulations' and 'you' to shout: yell
yell = shout('congratulations', 'you')

# Print yell
print(yell)

congratulations!!!you!!!


In [2]:
nums=(3,4,6)
num1,num2,num3=nums
print(num3)

6


- how to construct, unpack, and access tuple elements. 
- if we know a turple nums=(3,4,5)
- unpack this turple

In [19]:
#how to construct, unpack, and access tuple elements. 
#if we know a turple nums=(3,4,5)
#unpack this turple
nums=(3,4,6)
num1,num2,num3=nums

#change nums[0] into 2, 将第一个变量变成2，新的tuple
even_num= (2, num2,num3)
print(even_num)

(2, 4, 6)


In [3]:
#return multiple values from a function using tuples
#return x, y has the same result as return (x, y).the former actually packs x and y into a tuple under the hood
# Define shout_all with parameters word1 and word2
def shout_all(word1, word2):
    
    # Concatenate word1 with '!!!': shout1
    shout1= word1 + '!!!'
    
    # Concatenate word2 with '!!!': shout2
    shout2= word2 + '!!!'
    
    # Construct a tuple with shout1 and shout2: shout_words   重要 构建
    shout_words=(shout1, shout2)

    # Return shout_words
    return shout_words

# Pass 'congratulations' and 'you' to shout_all(): yell1, yell2  分解
yell1, yell2=shout_all('congratulations', 'you')
#这一步很重要：分解shout_all()其中的两个parameters，这样就可以单独查看
# Print yell1 and yell2
print(yell1)
print(yell2)

congratulations!!!
you!!!


In [ ]:
# how to load a dataset into a DataFrame
# The dataset contains Twitter data and you will iterate over entries in a column to build a dictionary in which the keys are the names of languages and the values are the number of tweets in the given language. The file tweets.csv 
# Import pandas
import pandas as pd
# Import Twitter data as DataFrame: df
df = pd.read_csv('tweets.csv')#read_csv is a pandas function

# Initialize an empty dictionary: langs_count
langs_count = {}

# Extract column from DataFrame: col
col = df['lang']

# Iterate over lang column in DataFrame
for entry in col:
#(To add 1 the value of a key entry in a dictionary d, use d[entry] += 1. To add a new key to d and set its value to 1, use d[entry] = 1)
    # If the language is in langs_count, add 1
    if entry in langs_count.keys():
     langs_count[entry]+=1
    # Else add the language to dictionary:langs_count, set the value to 1
    else: 
     langs_count[entry]=1

# Print the populated dictionary
print(langs_count)

In [ ]:
# Define count_entries()
def count_entries(df, col_name):
    """Return a dictionary with counts of 
    occurrences as value for each key."""

    # Initialize an empty dictionary: langs_count
    langs_count = {}
    
    # Extract column from DataFrame: col
    col = df[col_name]
    
    # Iterate over lang column in DataFrame
    for entry in col:

        # If the language is in langs_count, add 1
        if entry in langs_count.keys():
            langs_count[entry]+=1
        # Else add the language to langs_count, set the value to 1
        else:
            langs_count[entry]=1

    # Return the langs_count dictionary
    return langs_count
#注意加入return的两点 1.后面没有括号， 2.return是处于function之一级别的，不是for循环里面的内容，所以要与for对齐

# Call count_entries(): result
result=count_entries(tweets_df,'lang')

# Print the result
print(result)

# 学习三种scopes  
- 第一global scope。
- 第二local scope方程中有的variable，如果在方程外，不可以提取出来。
- 第三 build in scope

In [7]:
num=5
def func1(num):
    num = 3
    print(num)

func1(num)
   
#func1() prints out 3, func2() prints out 10, and the value of num in the global scope is 6.

3


In [14]:
num = 69
def func2():
    global num #注意这里是用的是global num 即是方程外定义的num
    double_num = num * 2
    num = 6#这里又改变的num
    print(double_num)
func2() 


138


In [16]:
# Create a string: team
team = "teen titans"

# Define change_team() 建立一个function
def change_team():
    """Change the value of the global variable team."""

    # Use team in global scope， global后面 空格加入
    global team

    # Change the value of team in global: team
    team="justice league"
# Print team
print(team)#他的结果是global的teen titans

# Call change_team()

change_team()#运用function，改变了global的，这里没有变量在空号内，所以运用时扩号里也没变量

# Print team
print(team)

teen titans
justice league


In [17]:
#check out Python's built-in scope
import builtins
dir(builtins)
这一块不太知道是什

SyntaxError: invalid syntax (<ipython-input-17-a68797b01ec1>, line 4)

# 建立function套function

In [18]:
#Nest functions  一个function 中 有 一个function
# Define three_shouts
def three_shouts(word1, word2, word3):
    """Returns a tuple of strings
    concatenated with '!!!'."""

    # Define inner
    def inner(word):
        """Returns a string concatenated with '!!!'."""
        return word + '!!!'

    # Return a tuple of strings 这里要注意，这里是外面的function的parameters在里面的inner function的运用
    return (inner(word1), inner(word2), inner(word3))
# Call three_shouts() and print
print(three_shouts('a', 'b', 'c'))

('a!!!', 'b!!!', 'c!!!')


# a lambda function
- """Concatenate echo copies of word1."""
- convert a function that does a simple task into a lambda function. 

In [20]:
#简单function
def echo_word(word1, echo):
    """Concatenate echo copies of word1."""
    words = word1 * echo
    return words

In [22]:
echo_word('hey', 5) 

'heyheyheyheyhey'

In [21]:
#转换为lambdafunction
# Define echo_word as a lambda function: echo_word
echo_word = (lambda word1, echo: word1 * echo)

# Call echo_word: result
result = echo_word('hey', 5)

# Print result
print(result)

heyheyheyheyhey


## use a lambda function--map() function.

- a lambda function, which raises a value a to the power of 2, 
- is passed to map() alongside a list of numbers, nums. 
- The map object that results from the call to map() is stored in result. 

In [24]:
nums = [2, 4, 6, 8, 10]

result = map(lambda a: a ** 2, nums)

print(result)

[4, 16, 36, 64, 100]


In [25]:
# Create a list of strings: spells
spells = ['protego', 'accio', 'expecto patronum', 'legilimens']

# Use map() to apply a lambda function over spells: shout_spells
shout_spells = map(lambda item: item + '!!!', spells)

# Convert shout_spells to a list: shout_spells_list
shout_spells_list = list(shout_spells)

# into a list and print it
print(shout_spells_list)

['protego!!!', 'accio!!!', 'expecto patronum!!!', 'legilimens!!!']


## Filter() and lambda functions

- **The function filter() offers a way to filter out elements from a list that don't satisfy certain criteria.**

In [26]:
# Create a list of strings: fellowship
fellowship = ['frodo', 'samwise', 'merry', 'aragorn', 'legolas', 'boromir', 'gimli']

# Use filter() to apply a lambda function over fellowship: result
result = filter(lambda member: len(member) > 6, fellowship)

# Convert result to a list: result_list
result_list = list(result)

# Convert result into a list and print it
print(result_list)

['samwise', 'aragorn', 'legolas', 'boromir']


## Reduce() and lambda functions

- returns a single value as a result. 

In [27]:
# Define gibberish
def gibberish(*args):
    """Concatenate strings in *args together."""
    hodgepodge = ''
    for word in args:
        hodgepodge += word
    return hodgepodge

In [28]:
# Import reduce from functools
from functools import reduce

In [29]:
# Create a list of strings: stark
stark = ['robb', 'sansa', 'arya', 'eddard', 'jon']

# Use reduce() to apply a lambda function over stark: result
result = reduce(lambda item1, item2: item1 + item2, stark)

# Print the result
print(result)

robbsansaaryaeddardjon


# Error handling with
- try-except


- raise. a raise statement to the shout_echo() function to raise an error message when the value supplied by the user to the echo argument is less than 0.


In [30]:
# Define shout_echo
def shout_echo(word1, echo=1):
    """Concatenate echo copies of word1 and three
    exclamation marks at the end of the string."""

    # Initialize empty strings: echo_word, shout_words
    echo_word = ''
    shout_words = ''

    # Add exception handling with try-except
    try:
        # Concatenate echo copies of word1 using *: echo_word
        echo_word = word1 * echo

        # Concatenate '!!!' to echo_word: shout_words
        shout_words = echo_word + '!!!'
    except:
        # Print error message
        print("word1 must be a string and echo must be an integer.")

    # Return shout_words
    return shout_words

# Call shout_echo
shout_echo("particle", echo="accelerator")


word1 must be a string and echo must be an integer.


''

In [35]:
# Define shout_echo
def shout_echo(word1, echo=1):
    """Concatenate echo copies of word1 and three
    exclamation marks at the end of the string."""

    # Raise an error with raise
    if echo < 0:
        raise ValueError('echo must be greater than 0')

    # Concatenate echo copies of word1 using *: echo_word
    echo_word = word1 * echo

    # Concatenate '!!!' to echo_word: shout_word
    shout_word = echo_word + '!!!'

    # Return shout_word
    return shout_word

# Call shout_echo
shout_echo("particle", echo=4)

'particleparticleparticleparticle!!!'

# write  functions using lambda, 
- how to pass lambda functions as arguments to other functions such as map(), filter(), and reduce(), 
- as well as how to write errors and output custom error messages within your functions. 
-  put together these learnings to good use by working with a Twitter dataset. 



1. Before practicing your new error handling skills,in this exercise, you will write a lambda function and use filter() to select retweets, that is, tweets that begin with the string 'RT'.


In [ ]:
# Select retweets from the Twitter DataFrame: result
result = filter(lambda x: x[0:2] == 'RT', tweets_df['text'])

# Create list from filter object result: res_list
res_list = list(result)

# Print all retweets in res_list
for tweet in res_list:
    print(tweet)
    
#The first argument to filter() should be the lambda function and 
#the second argument is the sequence of tweets, tweets_df['text'].
#To convert the filter object to a list, use the list() function.

 **improve on your previous work with the count_entries() function
 by adding a try-except block to it.**
 
 This will allow your function to provide a helpful message when the user calls your count_entries() function but provides a column name that isn't in the DataFrame.



In [ ]:
# Define count_entries()
def count_entries(df, col_name='lang'):
    """Return a dictionary with counts of
    occurrences as value for each key."""

    # Initialize an empty dictionary: cols_count
    cols_count = {}

    # Add try block
    try: 
        # Extract column from DataFrame: col
        col = df[col_name]
        
        # Iterate over the column in dataframe
        for entry in col:
    
            # If entry is in cols_count, add 1
            if entry in cols_count.keys():
                cols_count[entry] += 1
            # Else add the entry to cols_count, set the value to 1
            else:
                cols_count[entry] = 1
    
        # Return the cols_count dictionary
        return cols_count

    # Add except block
    except:
        print('The DataFrame does not have a ' + col_name + ' column.')

# Call count_entries(): result1
result1 = count_entries(tweets_df, 'lang')

# Print result1
print(result1)

# Call count_entries(): result2
#result2 = count_entries(tweets_df, 'lang1')

- **raise a ValueError in the case that the user provides a column name that isn't in the DataFrame.**

In [ ]:
# Define count_entries()
def count_entries(df, col_name='lang'):
    """Return a dictionary with counts of
    occurrences as value for each key."""
    
    # Raise a ValueError if col_name is NOT in DataFrame
    if col_name not in df.columns:
        raise ValueError('The DataFrame does not have a ' + col_name + ' column.')

    # Initialize an empty dictionary: cols_count
    cols_count = {}
    
    # Extract column from DataFrame: col
    col = df[col_name]
    
    # Iterate over the column in DataFrame
    for entry in col:

        # If entry is in cols_count, add 1
        if entry in cols_count.keys():
            cols_count[entry] += 1
            # Else add the entry to cols_count, set the value to 1
        else:
            cols_count[entry] = 1
        
        # Return the cols_count dictionary
    return cols_count

# Call count_entries(): result1
result1 = count_entries(tweets_df, 'lang')

# Print result1
print(result1)

# Iterators vs Iterables
-  an iterable is an object that can return an iterator, **while **an iterator is an object that keeps state and produces the next value when you call next() on it

- which object is an iterable and which is an iterator.



In [ ]:
In [1]: print(flash1)
['jay garrick', 'barry allen', 'wally west', 'bart allen']

In [2]: print(flash2)
<list_iterator object at 0x7fc024989208>

In [3]: next(flash2)
Out[3]: 'jay garrick'

In [4]: next(flash2)
Out[4]: 'barry allen'

In [5]: next(flash2)
Out[5]: 'wally west'

In [6]: next(flash2)
Out[6]: 'bart allen'

**flash1 is an iterable and flash2 is an iterator.**




In [38]:
# Create a list of strings: flash
flash = ['jay garrick', 'barry allen', 'wally west', 'bart allen']

# Print each list item in flash using a for loop
for person in flash:
    print(person)

jay garrick
barry allen
wally west
bart allen


In [39]:
# Create an iterator for flash: superspeed
superspeed = iter(flash)

# Print each item from the iterator
print(next(superspeed))
print(next(superspeed))
print(next(superspeed))
print(next(superspeed))

jay garrick
barry allen
wally west
bart allen


not all iterables are actual lists.

A couple of examples that we looked at are strings and the use of the range() function. In this exercise, we will focus on the range() function.

In [40]:
#use range() in a for loop as if it's a list to be iterated over:
for i in range(5):
    print(i)

0
1
2
3
4


In [45]:
# Create an iterator for range(3): small_value
small_value = iter(range(3))
type(small_value)

listiterator

In [42]:
# Print the values in small_value
print(next(small_value))
print(next(small_value))
print(next(small_value))

# Loop over range(3) and print the values
for num in range(3):
    print(num)

0
1
2
0
1
2


## Iterators as function arguments

-  using the iter() function to get an iterator object, as well as the next() function to retrieve the values one by one from the iterator object.

- also functions that take iterators as arguments. For example, the list() and sum() functions return a list and the sum of elements, respectively.

In [48]:
# Create a range object: values
values = range(10, 21)

# Print the range object
print(values)
type(values)

[10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]


list

In [49]:
## 转变成list

# Create a list of integers: values_list
values_list = list(values)

# Print values_list
print(values_list)
type(values_list)

[10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]


list

In [51]:
#对List求和-----sum()

# Get the sum of values: values_sum
values_sum = sum(values)

# Print values_sum
print(values_sum)
type(values_sum)

165


int

## Using enumerate
- enumerate() returns an enumerate object that produces a sequence of tuples, 
- **and each of the tuples is an index-value pair.**

In this exercise, you are given a list of strings mutants and you will practice using enumerate() on it by printing out a list of tuples and unpacking the tuples using a for loop.

In [59]:
# Create a list of strings: mutants
mutants = ['charles xavier', 
            'bobby drake', 
            'kurt wagner', 
            'max eisenhardt', 
            'kitty pride']

aliases = ['prof x', 'iceman', 'nightcrawler', 'magneto', 'shadowcat']
powers = ['telepathy',
          'thermokinesis',
          'teleportation',
         'magnetokinesis',
          'intangibility']

In [54]:
# Create a list of tuples: mutant_list
mutant_list = list(enumerate(mutants))
print(mutant_list)

[(0, 'charles xavier'), (1, 'bobby drake'), (2, 'kurt wagner'), (3, 'max eisenhardt'), (4, 'kitty pride')]


In [56]:
#(1)从0开始 unpack揭开
# Unpack and print the tuple pairs
for index1, value1 in enumerate(mutants):
    print(index1, value1)

#(2)从1开始 unpack揭开
# Change the start index ！！！！
for index2, value2 in enumerate(mutants, start=1):
    print(index2, value2)


(0, 'charles xavier')
(1, 'bobby drake')
(2, 'kurt wagner')
(3, 'max eisenhardt')
(4, 'kitty pride')
(1, 'charles xavier')
(2, 'bobby drake')
(3, 'kurt wagner')
(4, 'max eisenhardt')
(5, 'kitty pride')


## Using zip

- Another interesting function is zip(), which takes any number of iterables and returns a zip object that is an iterator of tuples.
- If you wanted to print the values of a zip object, you can convert it into a list and then print it. 

**Printing just a zip object will not return the values unless you unpack it first. **

Three lists of strings are pre-loaded: mutants, aliases, and powers. 
- 1.First, you will use list() and zip() on these lists to generate a list of tuples. 
- 2.Then, you will create a zip object using zip().
- 3.Finally, you will unpack this zip object in a for loop to print the values in each tuple. 

Observe the different output generated by printing the list of tuples, then the zip object, and finally, the tuple values in the for loop.

In [63]:
# Create a list of tuples: mutant_data
mutant_data = list(zip(mutants, aliases, powers))

# Print the list of tuples
print(mutant_data)


[('charles xavier', 'prof x', 'telepathy'), ('bobby drake', 'iceman', 'thermokinesis'), ('kurt wagner', 'nightcrawler', 'teleportation'), ('max eisenhardt', 'magneto', 'magnetokinesis'), ('kitty pride', 'shadowcat', 'intangibility')]


In [64]:
# Create a zip object using the three lists: mutant_zip
mutant_zip = zip(mutants, aliases, powers)

# Print the zip object
print(mutant_zip)


[('charles xavier', 'prof x', 'telepathy'), ('bobby drake', 'iceman', 'thermokinesis'), ('kurt wagner', 'nightcrawler', 'teleportation'), ('max eisenhardt', 'magneto', 'magnetokinesis'), ('kitty pride', 'shadowcat', 'intangibility')]


In [65]:
# Unpack the zip object and print the tuple values
for value1, value2, value3 in mutant_zip:
    print(value1, value2, value3)

('charles xavier', 'prof x', 'telepathy')
('bobby drake', 'iceman', 'thermokinesis')
('kurt wagner', 'nightcrawler', 'teleportation')
('max eisenhardt', 'magneto', 'magnetokinesis')
('kitty pride', 'shadowcat', 'intangibility')


**There is no unzip function for doing the reverse of what zip() does.**

We can, however, reverse what has been zipped together by using zip() with a little help from *! * unpacks an iterable such as a list or a tuple into positional arguments in a function call.

- use * in a call to zip() to unpack the tuples produced by zip().

- Two tuples of strings, mutants and powers have been pre-loaded.

In [70]:
# Create a zip object from mutants and powers: z1
z1 = zip(mutants, powers)

# Print the tuples in z1 by unpacking with *
print(*z1)

('charles xavier', 'telepathy') 
('bobby drake', 'thermokinesis') 
('kurt wagner', 'teleportation')
('max eisenhardt', 'magnetokinesis') 
('kitty pride', 'intangibility')

SyntaxError: invalid syntax (<ipython-input-70-66ac8d8ce5c1>, line 5)

In [67]:
# Re-create a zip object from mutants and powers: z1
z1 = zip(mutants, powers)

# 'Unzip' the tuples in z1 by unpacking with * and zip(): result1, result2
result1, result2 = zip(*z1)

# Check if unpacked tuples are equivalent to original tuples
print(result1 == mutants)
print(result2 == powers)

True
True

？？

False
False


# Using iterators to load bigdata


**Sometimes, the data we have to process reaches a size that is too much for a computer's memory to handle. This is a common problem faced by data scientists. **

**A solution to this is to process an entire data source chunk by chunk, instead of a single go all at once.**

You will process a large csv file of Twitter data in the same way that you processed 'tweets.csv' in Bringing it all together exercises of the prequel course, but this time, working on it in chunks of 10 entries at a time.

If you are interested in learning how to access Twitter data so you can work with it on your own system, refer to Part 2 of the DataCamp course on Importing Data in Python.


In [71]:
# Initialize an empty dictionary: counts_dict
counts_dict = {}

In [ ]:
# Iterate over the file chunk by chunk
for chunk in pd.read_csv('tweets.csv', chunksize=10):

    # Iterate over the column in DataFrame
    for entry in chunk['lang']:
        if entry in counts_dict.keys():
            counts_dict[entry] += 1
        else:
            counts_dict[entry] = 1

# Print the populated dictionary
print(counts_dict)


#<script.py> output:
#    {'en': 97, 'und': 2, 'et': 1}

## Extracting information --large amounts
- It's good to know how to process a file in smaller, more manageable chunks,
- but it can become very tedious having to write and rewrite the same code for the same task each time. 

- **making your code more reusable by putting your work in the last exercise in a function definition.**

The** pandas package has been imported as pd** and the file 'tweets.csv' is in your current directory for your use.


In [ ]:
# Define count_entries()
def count_entries(csv_file, c_size, colname):
    """Return a dictionary with counts of
    occurrences as value for each key."""
      
    # Initialize an empty dictionary: counts_dict  还是要建立一个dictionary
    counts_dict = {}

    # Iterate over the file chunk by chunk   这个其实就是上面的两个for loop
    for chunk in pd.read_csv(csv_file, chunksize=c_size):

        # Iterate over the column in DataFrame
        for entry in chunk[colname]:
            if entry in counts_dict.keys():    #结果要加入到dictionary[]中
                counts_dict[entry] += 1
            else:
                counts_dict[entry] = 1

    # Return counts_dict                        #最后要产生这个dictionary
    return counts_dict

In [ ]:
# Call count_entries(): result_counts
result_counts = count_entries('tweets.csv', 10, 'lang')

# Print result_counts
print(result_counts)

# List comprehensions
- collapes for loops into 1 single line 
- components : iterable , iterator variable , output expression
- 但是会牺牲 readability


## Write a basic list comprehension
writing list comprehensions. You will write a list comprehension and identify the output that will be produced.

doctor = ['house', 'cuddy', 'chase', 'thirteen', 'wilson']

How would a list comprehension that **produces a list of the first character of each string in doctor look like? Note that the list comprehension uses doc as the iterator variable. What will the output be?**


**The basic syntax for a list comprehension is [output expression for iterator variable in iterable].**

In [74]:
doctor = ['house', 'cuddy', 'chase', 'thirteen', 'wilson']

#doc as the iterator variable.
[doc[0] for doc in doctor]

['h', 'c', 'c', 't', 'w']

In [75]:
doctor = ['house', 'cuddy', 'chase', 'thirteen', 'wilson']

range(50)

underwood = 'After all, we are nothing more or less than what we choose to reveal.'

jean = '24601'

flash = ['jay garrick', 'barry allen', 'wally west', 'bart allen']

valjean = 24601

In [80]:
[doc[0] for doc in flash]

['j', 'b', 'w', 'b']

In [81]:
[doc[0] for doc in jean]

['2', '4', '6', '0', '1']

In [82]:
[doc[0] for doc in valjean]

TypeError: 'int' object is not iterable

In [87]:
[doc[0] for doc in range(50)]

TypeError: 'int' object has no attribute '__getitem__'

## list comprehension
- that produces a list of the squares of the numbers ranging from 0 to 9.


- MATRIX: by using nested listed comprehensions. 

Recall that you can create one of the rows of the matrix with a single list comprehension. 

To create the list of lists, you simply have to supply the list comprehension as the output expression of the overall list comprehension:

  - [[output expression] for iterator variable in iterable]

Note that here, the output expression is itself a list comprehension.

In [92]:
# Create list comprehension: squares
squares = [i**2 for i in range(0,10)]
print(squares)

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]


In [94]:
# Create a 5 x 5 matrix using a list of lists: matrix
matrix = [[col for col in range(5)] for row in range(5)]
matrix

[[0, 1, 2, 3, 4],
 [0, 1, 2, 3, 4],
 [0, 1, 2, 3, 4],
 [0, 1, 2, 3, 4],
 [0, 1, 2, 3, 4]]

In [95]:
for row in matrix:
    print(row)

[0, 1, 2, 3, 4]
[0, 1, 2, 3, 4]
[0, 1, 2, 3, 4]
[0, 1, 2, 3, 4]
[0, 1, 2, 3, 4]


# Conditional Advanced comprehensions
## **you can also create lists with values that meet only a certain condition. **

- One way of doing this is by using conditionals on iterator variables. 
Recall from the video that you can apply a conditional statement to test the iterator variable by adding an **if statement in the optional predicate expression part after the for statement in the comprehension:**

[ output expression** for iterator variable **in iterable **if** predicate expression ].

You will use this recipe to write a list comprehension for this exercise. 
You are given a list of strings fellowship and, using a list comprehension, you will create a list **that only includes the members of fellowship that have 7 characters or more.**

In [96]:
# Create a list of strings: fellowship
fellowship = ['frodo', 'samwise', 'merry', 'aragorn', 'legolas', 'boromir', 'gimli']

# Create list comprehension: new_fellowship
new_fellowship = [member for member in fellowship if len(member) >= 7]

# Print the new list
print(new_fellowship)

['samwise', 'aragorn', 'legolas', 'boromir']


## use an if-else statement on the output expression of the list.

You will work on the same list, fellowship and, using a list comprehension **and an if-else conditional** statement in the output expression, 
- create a list that keeps members of fellowship with 7 or more characters 
- and replaces others with an empty string. 
- Use member as the iterator variable in the list comprehension.

In [97]:
# Create a list of strings: fellowship
fellowship = ['frodo', 'samwise', 'merry', 'aragorn', 'legolas', 'boromir', 'gimli']

# Create list comprehension: new_fellowship
new_fellowship = [member if len(member) >= 7 else '' for member in fellowship]

# Print the new list
print(new_fellowship)

['', 'samwise', '', 'aragorn', 'legolas', 'boromir', '']


## Dict comprehensions

Comprehensions aren't relegated merely to the world of lists.
There are many other objects you can build using comprehensions, such as dictionaries, pervasive objects in Data Science. 
**You will create a dictionary using the comprehension syntax for this exercise. In this case, the comprehension is called a dict comprehension.**

Recall that the main difference between a list comprehension and a dict comprehension 
- **is the use of curly braces {} instead of []. **
- **Additionally, members of the dictionary are created using a colon :, as in key:value.**

You are given a list of strings fellowship and, using a dict comprehension, create a dictionary with the members of the list as the keys and the length of each string as the corresponding values.

In [98]:
# Create a list of strings: fellowship
fellowship = ['frodo', 'samwise', 'merry', 'aragorn', 'legolas', 'boromir', 'gimli']

# Create dict comprehension: new_fellowship
new_fellowship = {member:len(member) for member in fellowship}

# Print the new list
print(new_fellowship)

{'aragorn': 7, 'frodo': 5, 'samwise': 7, 'merry': 5, 'gimli': 5, 'boromir': 7, 'legolas': 7}


# generator expressions
- **A list comprehension produces a list as output, a generator produces a generator object.**
- the use of parentheses () in generator expressions and brackets [] in list comprehensions.
-  both be iterated over

In [6]:
# List of strings
fellowship = ['frodo', 'samwise', 'merry', 'aragorn', 'legolas', 'boromir', 'gimli']


In [10]:
# List comprehension
fellow1 = [member for member in fellowship if len(member) >= 7]

# Generator expression
fellow2 = (member for member in fellowship if len(member) >= 7)

print(fellow1)
fellow2

['samwise', 'aragorn', 'legolas', 'boromir']


<generator object <genexpr> at 0x10e038500>

## own generator expressions

In [28]:
result = (num for num in range(31))
result

<generator object <genexpr> at 0x10e0aae60>

In [29]:
# Print the first 5 values
print(next(result))
print(next(result))
print(next(result))
print(next(result))
print(next(result))

0
1
2
3
4


In [30]:
# Print the rest of the values
for value in result:
    print(value)

5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30


** GIVEN a list of strings lannister and, using a generator expression, create a generator object that you will iterate over to print its values.**


In [31]:
# Create a list of strings: lannister
lannister = ['cersei', 'jaime', 'tywin', 'tyrion', 'joffrey']

In [39]:
# Create a generator object: lengths
lengths = (len(person) for person in lannister)
print(lengths)

<generator object <genexpr> at 0x10e7eb410>


In [40]:
# Iterate over and print the values in lengths
for value in lengths:
    print(value)

6
5
5
6
7


## Generator functions 
- **are functions that, like generator expressions, yield a series of values, **
- **instead of returning a single value.**

A generator function is defined as you do a regular function, but whenever it generates a value,** it uses the keyword yield instead of return.**

- create a generator function with a similar mechanism as the generator expression you defined in the previous exercise:

In [41]:
# Define generator function get_lengths
def get_lengths(input_list):
    """Generator function that yields the
    length of the strings in input_list."""

    # Yield the length of a string
    for person in input_list:
        yield len(person)

# Print the values generated by get_lengths()
for value in get_lengths(lannister):
    print(value)


6
5
5
6
7


## List comprehensions for time-stamped data
- pandas Series,**a data structure that you will be working with a lot of times when analyzing data from pandas DataFrames. **
- **You can think of DataFrame columns as single-dimension arrays called Series**

 using a list comprehension to extract the time from time-stamped Twitter data. The pandas package has been imported as pd and the file 'tweets.csv' has been imported as the df DataFrame for your use.

In [ ]:
# Extract the created_at column from df: tweet_time
tweet_time = df['created_at']

# Extract the clock time: tweet_clock_time
tweet_clock_time = [entry[11:19] for entry in tweet_time]

# Print the extracted times
print(tweet_clock_time)

In [43]:
tweet_clock_time = ['23:40:17', '23:40:17', '23:40:17', '23:40:17', '23:40:17', '23:40:17', '23:40:18', '23:40:17', '23:40:18', '23:40:18', '23:40:18', '23:40:17', '23:40:18', '23:40:18', '23:40:17', '23:40:18', '23:40:18', '23:40:17', '23:40:18', '23:40:17', '23:40:18', '23:40:18', '23:40:18', '23:40:18', '23:40:17', '23:40:18', '23:40:18', '23:40:17', '23:40:18', '23:40:18', '23:40:18', '23:40:18', '23:40:18', '23:40:18', '23:40:18', '23:40:18', '23:40:18', '23:40:18', '23:40:18', '23:40:18', '23:40:18', '23:40:18', '23:40:18', '23:40:18', '23:40:18', '23:40:18', '23:40:18', '23:40:18', '23:40:18', '23:40:18', '23:40:18', '23:40:18', '23:40:18', '23:40:18', '23:40:18', '23:40:18', '23:40:18', '23:40:18', '23:40:18', '23:40:18', '23:40:19', '23:40:18', '23:40:18', '23:40:18', '23:40:19', '23:40:19', '23:40:19', '23:40:18', '23:40:19', '23:40:19', '23:40:19', '23:40:18', '23:40:19', '23:40:19', '23:40:19', '23:40:18', '23:40:19', '23:40:19', '23:40:19', '23:40:19', '23:40:19', '23:40:19', '23:40:19', '23:40:19', '23:40:19', '23:40:19', '23:40:19', '23:40:19', '23:40:19', '23:40:19', '23:40:19', '23:40:19', '23:40:19', '23:40:19', '23:40:19', '23:40:19', '23:40:19', '23:40:19', '23:40:19', '23:40:19']

In [ ]:
# Extract the created_at column from df: tweet_time
tweet_time = df['created_at']
#
#Each row is a string that represents a timestamp, 
#and you will access the 12th to 19th characters in the string to extract the time. 
#Use entry as the iterator variable and 
#assign the result to tweet_clock_time. 
#Additionally, add a conditional expression that checks whether entry[17:19] is equal to '19'.

# Extract the clock time: tweet_clock_time
tweet_clock_time = [entry[11:19] for entry in tweet_time if entry[17:19] == '19']

# Print the extracted times
print(tweet_clock_time)

# Case：建立字典

the World Bank's World Development Indicators dataset! You'll have the chance to write your own functions and list comprehensions as you work with iterators and generators and solidify your Python Data Science chops. Enjoy

Dictionaries for data science

- you'll use what you've learned about the zip() function and combine two lists into a dictionary.

- For pedagogical purposes, we have pre-processed this dataset into the lists that you'll be working with.

     - **The first list feature_names contains header names of the dataset 
     - and the second list row_vals contains actual values of a row from the dataset, corresponding to each of the header names.**

In [50]:
#the first argument you supply to zip() is feature_names, 
#followed by the second argument, row_vals.
feature_names= ['CountryName',
                'CountryCode',
                'IndicatorName',
                  'IndicatorCode',
                   'Year',
                   'Value']

In [51]:
row_vals =['Arab World',
       'ARB',
      'Adolescent fertility rate (births per 1,000 women ages 15-19)',
     'SP.ADO.TFRT',
     '1960',
      '133.56090740552298']

In [52]:
# Zip lists: zipped_lists
zipped_lists = zip(feature_names, row_vals)

# Create a dictionary: rs_dict  #need to pass the zip object zipped_lists to the call to dict().
rs_dict = dict(zipped_lists)

# Print the dictionary
print(rs_dict)

{'CountryCode': 'ARB', 'Value': '133.56090740552298', 'CountryName': 'Arab World', 'IndicatorName': 'Adolescent fertility rate (births per 1,000 women ages 15-19)', 'IndicatorCode': 'SP.ADO.TFRT', 'Year': '1960'}


Suppose you needed to **repeat the same process done in the previous exercise to many, many rows of data.**

Rewriting your code again and again could become very tedious, repetitive, and unmaintainable.

**create a function to house the code you wrote earlier to make things easier and much more concise.** Why? This way, you only need to call the function and supply the appropriate lists to create your dictionaries! 

Again, the lists feature_names and row_vals are preloaded and these contain the header names of the dataset and actual values of a row from the dataset, respectively.

In [53]:
# Define lists2dict()
def lists2dict(list1, list2):
    """Return a dictionary where list1 provides
    the keys and list2 provides the values."""

    # Zip lists: zipped_lists
    zipped_lists = zip(list1, list2)

    # Create a dictionary: rs_dict
    rs_dict = dict(zipped_lists)

    # Return the dictionary
    return rs_dict

# Call lists2dict: rs_fxn
rs_fxn = lists2dict(feature_names, row_vals)

# Print rs_fxn
print(rs_fxn)

{'CountryCode': 'ARB', 'Value': '133.56090740552298', 'CountryName': 'Arab World', 'IndicatorName': 'Adolescent fertility rate (births per 1,000 women ages 15-19)', 'IndicatorCode': 'SP.ADO.TFRT', 'Year': '1960'}


##  turn a bunch of lists into a list of dictionaries
**use the lists2dict() function you defined in the last exercise to turn a bunch of lists into a list of dictionaries with the help of a list comprehension.**


feature_names and row_lists. feature_names contains the header names of the World Bank dataset and row_lists is a list of lists, where each sublist is a list of actual values of a row from the dataset.


In [ ]:
# Print the first two lists in row_lists
print(row_lists[0])
print(row_lists[1])

# Turn list of lists into list of dicts: list_of_dicts
list_of_dicts = [lists2dict(feature_names, sublist) for sublist in row_lists]

# Print the first two dictionaries in list_of_dicts
print(list_of_dicts[0])
print(list_of_dicts[1])

## into a DataFrame

You've zipped lists together, created a function to house your code, and even used the function in a list comprehension to generate a list of dictionaries.

**You will now use of all these to convert the list of dictionaries into a pandas DataFrame. **You will see how convenient it is to generate a DataFrame from dictionaries with the DataFrame() function from the pandas package.


In [ ]:
# Import the pandas package
import pandas as pd

# Turn list of lists into list of dicts: list_of_dicts
list_of_dicts = [lists2dict(feature_names, sublist) for sublist in row_lists]


In [ ]:
#将整个字典转化为dataframe
# Turn list of dicts into a DataFrame: df
df = pd.DataFrame(list_of_dicts)

# Print the head of the DataFrame
print(df.head())


# Streaming Data
- generator function process one by one : chunk by chunk ; line by line 

- Processing data in chunks

     - Sometimes, data sources can be so large in size that storing the entire dataset in memory becomes too resource-intensive. 
     - process the first 1000 rows of a file line by line, to create a dictionary of the counts of how many times each country appears in a column in the dataset.

     - The csv file 'world_dev_ind.csv' is in your current directory for your use. 

     - To begin, you need to open a connection to this file using what is known as a context manager. 

For example, the command **with open('datacamp.csv')** as datacamp binds the csv file 'datacamp.csv' as datacamp in the context manager. Here, the **with statement** is the context manager, and its purpose is to ensure that resources are efficiently allocated when opening a connection to a file.

In [ ]:
# Open a connection to the file
with open('world_dev_ind.csv') as file:

    # Skip the column names
    file.readline()

    # Initialize an empty dictionary: counts_dict
    counts_dict = {}

    # Process only the first 1000 rows
    for j in range(0, 1000):

        # Split the current line into a list: line
        line = file.readline().split(',')

        # Get the value for the first column: first_col
        first_col = line[0]

        # If the column value is in the dict, increment its value
        if first_col in counts_dict.keys():
            counts_dict[first_col] += 1

        # Else, add to the dict and set value to 1
        else:
            counts_dict[first_col] = 1

# Print the resulting dictionary
print(counts_dict)


output:

    {'Fragile and conflict affected situations': 76, 'Europe & Central Asia (all income levels)': 109, 'Europe & Central Asia (developing only)': 89, 'Arab World': 80, 'European Union': 116, 'Heavily indebted poor countries (HIPC)': 18, 'Caribbean small states': 77, 'East Asia & Pacific (all income levels)': 122, 'Euro area': 119, 'Central Europe and the Baltics': 71, 'East Asia & Pacific (developing only)': 123}


 What if, however, you want to do this for the entire file?
## Generators

**Generators allow users to lazily evaluate data. This concept of lazy evaluation is useful when you have to deal with very large datasets because it lets you generate values in an efficient manner by yielding only chunks of data at a time instead of the whole thing at once.**

- define a generator function read_large_file() that produces a generator object 
- which yields a single line from a file 
- each time next() is called on it.


  - **Note that when you open a connection to a file, the resulting file object is already a generator!** So out in the wild, you won't have to explicitly create generator objects in cases such as this. 
  - However, for pedagogical reasons, we are having you practice how to do this here with the read_large_file() function.

In [ ]:
# Define read_large_file()
def read_large_file(file_object):
    """A generator function to read a large file lazily."""

    # Loop indefinitely until the end of the file
    while True:

        # Read a line from the file: data
        data = file_object.readline()

        # Break if this is the end of the file
        if not data:
            break

        # Yield the line of data
        yield data

# Open a connection to the file
with open('world_dev_ind.csv') as file:

    # Create a generator object for the file: gen_file
    gen_file = read_large_file(file)

    # Print the first three lines of the file
    print(next(gen_file))
    print(next(gen_file))
    print(next(gen_file))


output:
    CountryName,CountryCode,IndicatorName,IndicatorCode,Year,Value
    
    Arab World,ARB,"Adolescent fertility rate (births per 1,000 women ages 15-19)",SP.ADO.TFRT,1960,133.56090740552298
    
    Arab World,ARB,Age dependency ratio (% of working-age population),SP.POP.DPND,1960,87.7976011532547
    

**You will process the file line by line, to create a dictionary of the counts of how many times each country appears in a column in the dataset. **

For this exercise, however, you won't process just 1000 rows of data, you'll process the entire dataset!

In [ ]:
# Initialize an empty dictionary: counts_dict
counts_dict = {}

# Open a connection to the file
with open('world_dev_ind.csv') as file:

    # Iterate over the generator from read_large_file()
    for line in read_large_file(file):

        row = line.split(',')
        first_col = row[0]

        if first_col in counts_dict.keys():
            counts_dict[first_col] += 1
        else:
            counts_dict[first_col] = 1

# Print            
print(counts_dict)

   {'High income: nonOECD': 68, 'Europe & Central Asia (developing only)': 89, 'Least developed countries: UN classification': 78, 'South Asia': 36, 'Caribbean small states': 77, 'Middle East & North Africa (developing only)': 94, 'Europe & Central Asia (all income levels)': 109, 'Fragile and conflict affected situations': 76, 'Lower middle income': 126, 'CountryName': 1, 'Euro area': 119, 'Small states': 69, 'Arab World': 80, 'Latin America & Caribbean (developing only)': 133, 'Central Europe and the Baltics': 71, 'OECD members': 130, 'Middle income': 138, 'High income': 131, 'Middle East & North Africa (all income levels)': 89, 'European Union': 116, 'Latin America & Caribbean (all income levels)': 130, 'Low & middle income': 138, 'North America': 123, 'Pacific island small states': 66, 'Heavily indebted poor countries (HIPC)': 99, 'East Asia & Pacific (all income levels)': 122, 'High income: OECD': 127, 'Other small states': 63, 'East Asia & Pacific (developing only)': 123, 'Low income': 80}
   
   
 # Using pandas read_csv iterator 
**Another way to read data too large to store in memory in chunks is to read the file in as DataFrames of a certain length, say, 100.**
- For example, with the pandas package (imported as pd), **you can do pd.read_csv(filename, chunksize=100). &&
- This creates an iterable reader object, which means that you can use next() on it.


In [ ]:
# Import the pandas package
import pandas as pd

# Initialize reader object: df_reader
df_reader = pd.read_csv('ind_pop.csv', chunksize=10)

# Print two chunks !!!!!
print(next(df_reader))
print(next(df_reader))

#### Chunk 1
CountryName CountryCode  \
0                                 Arab World         ARB   
1                     Caribbean small states         CSS   
2             Central Europe and the Baltics         CEB   
3    East Asia & Pacific (all income levels)         EAS   
4      East Asia & Pacific (developing only)         EAP   
5                                  Euro area         EMU   
6  Europe & Central Asia (all income levels)         ECS   
7    Europe & Central Asia (developing only)         ECA   
8                             European Union         EUU   
9   Fragile and conflict affected situations         FCS   

                   IndicatorName      IndicatorCode  Year      Value  
0  Urban population (% of total)  SP.URB.TOTL.IN.ZS  1960  31.285384  
1  Urban population (% of total)  SP.URB.TOTL.IN.ZS  1960  31.597490  
2  Urban population (% of total)  SP.URB.TOTL.IN.ZS  1960  44.507921  
3  Urban population (% of total)  SP.URB.TOTL.IN.ZS  1960  22.471132  
4  Urban population (% of total)  SP.URB.TOTL.IN.ZS  1960  16.917679  
5  Urban population (% of total)  SP.URB.TOTL.IN.ZS  1960  62.096947  
6  Urban population (% of total)  SP.URB.TOTL.IN.ZS  1960  55.378977  
7  Urban population (% of total)  SP.URB.TOTL.IN.ZS  1960  38.066129  
8  Urban population (% of total)  SP.URB.TOTL.IN.ZS  1960  61.212898  
9  Urban population (% of total)  SP.URB.TOTL.IN.ZS  1960  17.891972

#### Chunk 2 
                                      CountryName CountryCode  \
10         Heavily indebted poor countries (HIPC)         HPC   
11                                    High income         HIC   
12                           High income: nonOECD         NOC   
13                              High income: OECD         OEC   
14  Latin America & Caribbean (all income levels)         LCN   
15    Latin America & Caribbean (developing only)         LAC   
16   Least developed countries: UN classification         LDC   
17                            Low & middle income         LMY   
18                                     Low income         LIC   
19                            Lower middle income         LMC   

                    IndicatorName      IndicatorCode  Year      Value  
10  Urban population (% of total)  SP.URB.TOTL.IN.ZS  1960  12.236046  
11  Urban population (% of total)  SP.URB.TOTL.IN.ZS  1960  62.680332  
12  Urban population (% of total)  SP.URB.TOTL.IN.ZS  1960  56.107863  
13  Urban population (% of total)  SP.URB.TOTL.IN.ZS  1960  64.285435  
14  Urban population (% of total)  SP.URB.TOTL.IN.ZS  1960  49.284688  
15  Urban population (% of total)  SP.URB.TOTL.IN.ZS  1960  44.863308  
16  Urban population (% of total)  SP.URB.TOTL.IN.ZS  1960   9.616261  
17  Urban population (% of total)  SP.URB.TOTL.IN.ZS  1960  21.272894  
18  Urban population (% of total)  SP.URB.TOTL.IN.ZS  1960  11.498396  
19  Urban population (% of total)  SP.URB.TOTL.IN.ZS  1960  19.810513

**using a bigger DataFrame chunk size and then process the data from the first chunk.**

1. To process the data, you will create another DataFrame composed of only the rows from a specific country. 
2. You will then zip together two of the columns from the new DataFrame, 'Total Population' and 'Urban population (% of total)'.
3. Finally, you will create a list of tuples from the zip object, where each tuple is composed of a value from each of the two columns mentioned.


In [ ]:
# Initialize reader object: urb_pop_reader
urb_pop_reader = pd.read_csv('ind_pop_data.csv', chunksize=1000)

# Get the first DataFrame chunk: df_urb_pop
df_urb_pop = next(urb_pop_reader)

# Check out the head of the DataFrame
print(df_urb_pop.head())

# Check out specific country: df_pop_ceb
df_pop_ceb = df_urb_pop[df_urb_pop['CountryCode'] == 'CEB']


                               CountryName CountryCode  Year  \
0                               Arab World         ARB  1960   
1                   Caribbean small states         CSS  1960   
2           Central Europe and the Baltics         CEB  1960   
3  East Asia & Pacific (all income levels)         EAS  1960   
4    East Asia & Pacific (developing only)         EAP  1960   

   Total Population  Urban population (% of total)  
0      9.249590e+07                      31.285384  
1      4.190810e+06                      31.597490  
2      9.140158e+07                      44.507921  
3      1.042475e+09                      22.471132  
4      8.964930e+08                      16.917679

**df_pop_ceb = df_urb_pop[df_urb_pop['CountryCode'] == 'CEB']**

                        CountryName CountryCode  Year  Total Population  \
2    Central Europe and the Baltics         CEB  1960        91401583.0   
244  Central Europe and the Baltics         CEB  1961        92237118.0   
486  Central Europe and the Baltics         CEB  1962        93014890.0   
728  Central Europe and the Baltics         CEB  1963        93845749.0   
970  Central Europe and the Baltics         CEB  1964        94722599.0   

     Urban population (% of total)  
2                        44.507921  
244                      45.206665  
486                      45.866565  
728                      46.534093  
970                      47.208743



In [ ]:
# Zip DataFrame columns of interest: pops
pops = zip(df_pop_ceb['Total Population'], 
            df_pop_ceb['Urban population (% of total)'])

# Turn zip object into list: pops_list
pops_list = list(pops)

# Print pops_list
print(pops_list)

[(91401583.0, 44.507921139002597), (92237118.0, 45.206665319194002), (93014890.0, 45.866564696018003), (93845749.0, 46.5340927663649), (94722599.0, 47.208742980352604)]

##  adding a column to a DataFrame.
**using a list comprehension to create the values for a new column 'Total Urban Population' from the list of tuples that you generated earlier. **

- The values in this new column 'Total Urban Population', therefore, are the product of the first and second element in each tuple. 
- Furthermore, because the 2nd element is a percentage, you need to divide the entire result by 100, or alternatively, multiply it by 0.01.

- plot the data from this new column to create a visualization of the urban population data.



In [ ]:
# Use list comprehension to create new DataFrame column 'Total Urban Population'
df_pop_ceb['Total Urban Population'] = [int(tup[0] * tup[1] * 0.01) for tup in pops_list]

# Plot urban population data
df_pop_ceb.plot(kind='scatter', x='Year', y='Total Urban Population')
plt.show()

## process the entire large dataset by just working on smaller pieces of it!

In [ ]:
# Initialize reader object: urb_pop_reader
urb_pop_reader = pd.read_csv('ind_pop_data.csv', chunksize=1000)

# Initialize empty DataFrame: data
data = pd.DataFrame()

# Iterate over each DataFrame chunk
for df_urb_pop in urb_pop_reader:

    # Check out specific country: df_pop_ceb
    df_pop_ceb = df_urb_pop[df_urb_pop['CountryCode'] == 'CEB']

    # Zip DataFrame columns of interest: pops
    pops = zip(df_pop_ceb['Total Population'],
                df_pop_ceb['Urban population (% of total)'])

    # Turn zip object into list: pops_list
    pops_list = list(pops)

    # Use list comprehension to create new DataFrame column 'Total Urban Population'
    df_pop_ceb['Total Urban Population'] = [int(tup[0] * tup[1]) for tup in pops_list]
    
    # Append DataFrame chunk to data: data
    data = data.append(df_pop_ceb)

# Plot urban population data
data.plot(kind='scatter', x='Year', y='Total Urban Population')
plt.show()

#  put all the code for processing the data into a single function 
**so that you can reuse the code without having to rewrite the same things all over again.**

You're going to define the function plot_pop() which takes two arguments: 
- the filename of the file to be processed, 
- and the country code of the rows you want to process in the dataset.

Because all of the previous code you've written in the previous exercises will be housed in **plot_pop(),** calling the function already does the following:

1. Loading of the file chunk by chunk,
2. Creating the new column of urban population values, and
3. Plotting the urban population data.


You're going to use the data from 'ind_pop_data.csv', available in your current directory. The packages pandas and matplotlib.pyplot has been imported as pd and plt respectively for your use.

After you are done, take a moment to look at the plots and reflect on the new skills you have acquired. The journey doesn't end here! 


In [ ]:
# Define plot_pop()
def plot_pop(filename, country_code):

    # Initialize reader object: urb_pop_reader
    urb_pop_reader = pd.read_csv(filename, chunksize=1000)

    # Initialize empty DataFrame: data
    data = pd.DataFrame()
    
    # Iterate over each DataFrame chunk
    for df_urb_pop in urb_pop_reader:
        # Check out specific country: df_pop_ceb
        df_pop_ceb = df_urb_pop[df_urb_pop['CountryCode'] == country_code]

        # Zip DataFrame columns of interest: pops
        pops = zip(df_pop_ceb['Total Population'],
                    df_pop_ceb['Urban population (% of total)'])

        # Turn zip object into list: pops_list
        pops_list = list(pops)

        # Use list comprehension to create new DataFrame column 'Total Urban Population'
        df_pop_ceb['Total Urban Population'] = [int(tup[0] * tup[1]) for tup in pops_list]
    
        # Append DataFrame chunk to data: data
        data = data.append(df_pop_ceb)

    # Plot urban population data
    data.plot(kind='scatter', x='Year', y='Total Urban Population')
    plt.show()

# Set the filename: fn
fn = 'ind_pop_data.csv'

# Call plot_pop for country code 'CEB'
plot_pop(fn, 'CEB')

# Call plot_pop for country code 'ARB'
plot_pop(fn, 'ARB')
